### PR TITLE
Proposed lint.xml config file for Android

### PR DESCRIPTION
**Pull request by mpeng3**
Attached is the complete list of options available (reported via lint --list).  This proposed lint.xml file only includes the differences from the Android Studio defaults.

[lint options.txt](https://github.com/IntrepidPursuits/sherpa/files/905862/lint.options.txt)

### CODE REVIEW

File: android/lint.xml [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    <issue id="GoogleAppIndexingApiWarning" severity="warning"/>
+
+    <!-- Enable checking for app links -->
+    <issue id="AppLinksAutoVerifyError" severity="warning"/>
+
+    <!-- Enable checking for potential app links -->
+    <issue id="AppLinksAutoVerifyWarning" severity="warning"/>
+
+    <!-- Enable weak checking for UI design back button issues -->
+    <issue id="BackButton" severity="weak warning"/>
```
Comment (gdaniels) [inline android/lint.xml]: Meh.  I think we can do without this one?

File: android/lint.xml [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    <issue id="Registered" severity="error"/>
+
+    <!-- Upgrade severity of incorrect width/height values for ScrollView children -->
+    <issue id="ScrollViewSize" severity="error"/>
+
+    <!-- Upgrade severity of explicit path references to /sdcard -->
+    <issue id="SdCardPath" severity="error"/>
+
+    <!-- Enable checking for insecure random number generation -->
+    <issue id="SecureRandom" severity="weak warning"/>
```
Comment (gdaniels) [inline android/lint.xml]: this should be an error by default, IMO - then if people want to disable it for testing that's fine.

File: android/lint.xml [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    <issue id="GetInstance" severity="error"/>
+
+    <!-- Check for easter eggs -->
+    <issue id="EasterEgg" severity="weak warning"/>
+
+    <!-- Look for STOPSHIP comments -->
+    <issue id="StopShip" severity="error"/>
+
+    <!-- Look for mangled file line endings -->
+    <issue id="MangledCRLF" severity="weak warning"/>
```
Comment (f2rf2r) [inline android/lint.xml]: we might want to upgrade this for so files are consistent

File: android/lint.xml [hunk 4]
```diff
@@ -0,0 +1,139 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<lint>
+    <!-- list of issues to configure -->
```
Comment (f2rf2r) [inline android/lint.xml]: add a link to http://tools.android.com/tips/lint-checks so people can quickly reference the issues

**Review by f2rf2r — COMMENTED** [review]
Given that we are moving towards a 0 lint warning goal, we should turn most of these weak warnings to ignored or info

**Comment by f2rf2r** [comment]
Also, this might be a bit controversial, but should we also ignore the `ContentDescription` warnings? While it is sometimes helpful for people with accessibility needs, it's also very tedious to come up with a description for every image, and sometimes the descriptions might not be apparent and we would need to consult with the designers/pm/etc which might drag on the development time.

**Comment by gsintrepid** [comment]
@f2rf2r My two unexperienced cents:
I think it's good practice to advocate for accessibility and leave the `ContentDescription` warning by default. If a particular project decides on suppressing it, that seems like a decision the product owner should own.
Also it's often on icons or images that provide no content and the warning should be suppressed by definition anyway...

**Comment by gsintrepid** [comment]
Also small suggestions, but how about:

- Adding how-to instructions and a link to them in the Readme? The PR would be longer but at least those wouldn't be left behind before people start using the file.

- Using the same name as [skeleton's lint config file](https://github.com/IntrepidPursuits/AndroidSkeleton/blob/develop/app/lintConfig.xml), and also adding the same issue for skeleton dependencies? Or maybe it's the other way around and after this gets approved the skeleton would update its lint config file?

**Comment by f2rf2r** [comment]
I'm also planning on adding this to the [Static Analysis Plugin](https://github.com/IntrepidPursuits/static-analysis-gradle-plugin/) since they perform similar jobs

File: android/lint.xml [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+    <issue id="PermissionImpliesUnsupportedChromeOsHardware" severity="warning"/>
+
+    <!-- Enable checking for unsupported Chrome OS hardware -->
+    <issue id="UnsupportedChromeOsHardware" severity="error"/>
+
+    <!-- Enable checking for insecure cipher mode -->
+    <issue id="GetInstance" severity="error"/>
+
+    <!-- Check for easter eggs -->
+    <issue id="EasterEgg" severity="weak warning"/>
```
Comment (gdaniels) [inline android/lint.xml]: Pretty ambivalent on this one.  What if we want to ship an easter egg some day? :)

File: android/lint.xml [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+    <issue id="IconExpectedSize" severity="warning"/>
+
+    <!-- Enable WebP icon conversion suggestion -->
+    <issue id="ConvertToWebp" severity="info"/>
+
+    <!-- Upgrade severity of this check to improve layout computation speed -->
+    <issue id="DisableBaselineAlignment" severity="error"/>
+
+    <!-- Add support for checking for logging calls in production code -->
+    <issue id="LogConditional" severity="weak warning"/>
```
Comment (gdaniels) [inline android/lint.xml]: I think this is good behavior to encourage, but this is going to be a rough one, as I think we typically leave a lot of these in.

File: android/lint.xml [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+    <issue id="IllegalResourceRef" severity="error"/>
+
+    <!-- Upgrade severity of subtle manifest bugs -->
+    <issue id="ManifestOrder" severity="error"/>
+
+    <!-- Upgrade severity of missing minimum and target SDK attributes -->
+    <issue id="UsesMinSdkAttributes" severity="error"/>
+
+    <!-- Enable warning for negative margins -->
+    <issue id="NegativeMargin" severity="weak warning"/>
```
Comment (gdaniels) [inline android/lint.xml]: -1 (haha).  These aren't so bad, and we have some situations where we've used them effectively.

**Comment by f2rf2r** [comment]
We should also add the exclude for the InvalidPackage error of common 3rd party libs. ex:
```   
 <issue id="InvalidPackage">
        <ignore regexp="okio-.*jar"/>
        <ignore regexp="retrofit-.*jar"/>
        <ignore regexp="requery-.*jar"/>
    </issue>
```

**Comment by deanshi** [comment]
Are there any thoughts on the "UnusedAttribute" lint warning? I know specifically for us we have ~293 warnings just from using attributes that exist in API 24 (especially in XML vectors using fillType="evenOdd") but not below.

**Comment by deanshi** [comment]
Also thoughts on the "NewApi" lint errors similar to above? It seems that `android:letterSpacing` is a pretty big problem under this category.

**Comment by deanshi** [comment]
So there is a way to define these in separate, API specific styles (http://blog.octo.com/en/android-themes-styles-a-real-architecture/). This might be a good alternative rather than disabling/suppressing it.

File: android/lint.xml (hunk 6 above)
Comment (mpeng3) [reply android/lint.xml]: @gdaniels What do you think we should do about this one?  Leave it as a weak warning, or eliminate the check for now?

**Comment by mpeng3** [comment]
@deanshi Sorry for the supremely long delay in responding to this.  It sounds like leaving the UnusedAttribute check in place is best, but what about the "NewApi" one?  This seems like a good check to leave in, and to set the target api on a case-by-case basis as needed to indicate that this is the intended behavior.

**Comment by mpeng3** [comment]
@gsintrepid I believe we'll need to keep the name of this file as "lint.xml" in order for it to work as intended (being used by Android Studio's runtime lint checker).  With regards to the README file, do you see us needing anything more in-depth than the information provided on this page? https://developer.android.com/studio/write/lint.html

**Comment by gsintrepid** [comment]
@mpeng3 So does "Inspect Code" work when the file is named `lint.xml`?
That would be a nice thing to mention in the README (the main one or a sub-article).
(And maybe we could rename [`default-lintConfig.xml`](https://github.com/IntrepidPursuits/static-analysis-gradle-plugin/blob/master/src/main/resources/default-lintConfig.xml) to `default-lint.xml` in the Static Analysis Gradle Plugin, @f2rf2r?)

Also [`android_jenkins.md`](https://github.com/IntrepidPursuits/sherpa/blob/f5abb542c7ca3805bb12b72a907be43ba7f3392e/android/android_jenkins.md#additional-instructions-for-android-template) mentions that the Static Analysis Gradle Plugin is required by the default gradle configuration.
The README could explain that the lint file is the same used in the plugin (I assume they'll be the same), and if for some reason the project doesn't use the plugin, where to add the standalone lint file.

And even though the list of lint checks is in the file itself, it wouldn't hurt to add the link to it, and the link you just sent:
- https://developer.android.com/studio/write/lint.html
- http://tools.android.com/tips/lint-checks

File: android/android_project_creation.md [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+2. Run the Gradle task to analyze the code, which will generate the lint.xml file in the correct location in your project: `./gradlew analyze${buildVariant}` (e.g. `analyzeDevDebug`)
+
+3. Add this lint.xml file to Git and check it in.
+
+That's it!  Your project should now be configured to check against the recommended lint inspections.  You may modify this lint.xml file if needed to suit your project, but this should only be considered in exceptional circumstances.
+
+<b>Note:</b> If you ever want to update the lint.xml file to the latest version, do the following:
+1. Upgrade to the newest version of the `static-analysis-gradle-plugin`
+2. Delete the `lint.xml` file
```
Comment (f2rf2r) [inline android/android_project_creation.md]: include a note on why this is necessary

**Review by f2rf2r — APPROVED** [review]

**Pull request merged**